src/persist-client/src/internal/trace.rs

```diff
@@ -1113,6 +1137,27 @@ impl<T: Timestamp + Lattice> Spine<T> {
         }
     }
 
+    /// Apply some amount of effort to trace maintenance.
+    ///
+    /// The units of effort are updates, and the method should be thought of as
+    /// analogous to inserting as many empty updates, where the trace is
+    /// permitted to perform proportionate work.
+    ///
+    /// Returns true if this did work and false if it left the spine unchanged.
+    fn exert(&mut self, effort: &mut isize, log: &mut SpineLog<'_, T>) -> bool {
+        // If any merges exist, we can directly call `apply_fuel`.
+        if self.merging.iter().any(|b| b.merge.is_some()) {
+            self.apply_fuel(effort, log);
+            true
+        } else {
+            // TODO: Should we resurrect the rest of the old Spine::exert
```

Comment:
Where did we land on the empty batch thing [in the thread](https://materializeinc.slack.com/archives/C03P1EGB7A7/p1725464877346929)? I'm re-reading it and we decided how to do it, but not sure if we decided it was worth it or not

Reply:
I don't feel strongly! I suppose we could check a few catalog shards and see whether there was an incompletely-fuelled merge for the insertions and retractions or not, if we want to make a data-driven decision. (If ~all the big shards have such a merge then this is a nice conservative change; if not then we're going to need to do the empty-batch thing and might as well frontload it.)

Reply:
Here's the flat trace before/after that diff

```
FlatTrace { since: Antichain { elements: [0] }, legacy_batches: {
  HollowBatch { desc: ([0], [1], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/pee878e79-a060-4947-830c-1a8a79258f23"), encoded_size_bytes: 896, key_lower: [48], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([1], [2], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/pd462264c-79b5-4785-b927-e6896b5a20cc"), encoded_size_bytes: 896, key_lower: [49], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([2], [3], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/p5e37fba1-aa9f-4a6b-8dc3-21762b1a1ddd"), encoded_size_bytes: 896, key_lower: [50], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([3], [4], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/pff2a7171-4246-425f-ae42-82368cfdb302"), encoded_size_bytes: 896, key_lower: [51], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([4], [5], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/p6c16a144-612a-4d47-a080-91d3b6eb58a6"), encoded_size_bytes: 900, key_lower: [52], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([5], [5], [0]), parts: [], len: 0, runs: [], run_meta: [] }: ()
}, hollow_batches: {}, spine_batches: {
  SpineId(0, 4): ThinSpineBatch { level: 2, desc: Description { lower: Antichain { elements: [0] }, upper: Antichain { elements: [4] }, since: Antichain { elements: [0] } }, parts: [SpineId(0, 1), SpineId(1, 2), SpineId(2, 3), SpineId(3, 4)], descs: [Description { lower: Antichain { elements: [0] }, upper: Antichain { elements: [1] }, since: Antichain { elements: [0] } }, Description { lower: Antichain { elements: [1] }, upper: Antichain { elements: [2] }, since: Antichain { elements: [0] } }, Description { lower: Antichain { elements: [2] }, upper: Antichain { elements: [3] }, since: Antichain { elements: [0] } }, Description { lower: Antichain { elements: [3] }, upper: Antichain { elements: [4] }, since: Antichain { elements: [0] } }] },
  SpineId(4, 6): ThinSpineBatch { level: 1, desc: Description { lower: Antichain { elements: [4] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }, parts: [SpineId(4, 6)], descs: [Description { lower: Antichain { elements: [4] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }] },
  SpineId(6, 7): ThinSpineBatch { level: 0, desc: Description { lower: Antichain { elements: [5] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }, parts: [SpineId(6, 7)], descs: [Description { lower: Antichain { elements: [5] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }] }
}, merges: {} }
```

```
FlatTrace { since: Antichain { elements: [0] }, legacy_batches: {
  HollowBatch { desc: ([0], [1], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/pee878e79-a060-4947-830c-1a8a79258f23"), encoded_size_bytes: 896, key_lower: [48], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([1], [2], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/pd462264c-79b5-4785-b927-e6896b5a20cc"), encoded_size_bytes: 896, key_lower: [49], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([2], [3], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/p5e37fba1-aa9f-4a6b-8dc3-21762b1a1ddd"), encoded_size_bytes: 896, key_lower: [50], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([3], [4], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/pff2a7171-4246-425f-ae42-82368cfdb302"), encoded_size_bytes: 896, key_lower: [51], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([4], [5], [0]), parts: [Hollow(HollowBatchPart { key: PartialBatchKey("n0000000/p6c16a144-612a-4d47-a080-91d3b6eb58a6"), encoded_size_bytes: 900, key_lower: [52], stats: Some(LazyPartStats { key: ProtoStructStats { len: 1, cols: {"": ProtoDynStats { option: None, kind: Some(None(())) }} } }), ts_rewrite: None, diffs_sum: Some([1, 0, 0, 0, 0, 0, 0, 0]), format: None, schema_id: Some(SchemaId(0)) })], len: 1, runs: [], run_meta: [RunMeta { order: None, schema: None }] }: (),
  HollowBatch { desc: ([5], [5], [0]), parts: [], len: 0, runs: [], run_meta: [] }: ()
}, hollow_batches: {}, spine_batches: {
  SpineId(0, 4): ThinSpineBatch { level: 2, desc: Description { lower: Antichain { elements: [0] }, upper: Antichain { elements: [4] }, since: Antichain { elements: [0] } }, parts: [SpineId(0, 1), SpineId(1, 2), SpineId(2, 3), SpineId(3, 4)], descs: [Description { lower: Antichain { elements: [0] }, upper: Antichain { elements: [1] }, since: Antichain { elements: [0] } }, Description { lower: Antichain { elements: [1] }, upper: Antichain { elements: [2] }, since: Antichain { elements: [0] } }, Description { lower: Antichain { elements: [2] }, upper: Antichain { elements: [3] }, since: Antichain { elements: [0] } }, Description { lower: Antichain { elements: [3] }, upper: Antichain { elements: [4] }, since: Antichain { elements: [0] } }] },
  SpineId(4, 6): ThinSpineBatch { level: 1, desc: Description { lower: Antichain { elements: [4] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }, parts: [SpineId(4, 6)], descs: [Description { lower: Antichain { elements: [4] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }] },
  SpineId(6, 7): ThinSpineBatch { level: 0, desc: Description { lower: Antichain { elements: [5] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }, parts: [SpineId(6, 7)], descs: [Description { lower: Antichain { elements: [5] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }] },
  SpineId(7, 8): ThinSpineBatch { level: 0, desc: Description { lower: Antichain { elements: [5] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }, parts: [SpineId(7, 8)], descs: [Description { lower: Antichain { elements: [5] }, upper: Antichain { elements: [5] }, since: Antichain { elements: [0] } }] }
}, merges: {SpineId(6, 8): ThinMerge { since: Antichain { elements: [0] }, remaining_work: 0, active_compaction: None }} }
```

Reply:
Any idea how you ended up with `HollowBatch { desc: ([5], [5], [0]), ... }`?

A glance at the code suggests you're creating an empty spine batch and not that kind of trivial hollow batch, which I'd thought was disallowed somewhere anyways... assuming it's not clear to you how it appears, I'd guess that something about the legacy-batch-migration code is misbehaving? 🤔

Reply:
> Any idea how you ended up with HollowBatch { desc: ([5], [5], [0]), ... }?

Oh I explicitly made that. I thought that's what you suggested in the slack thread but apparently I misunderstood!

Reply:
Ah, got it!

In that case I'm pretty confident this is the legacy batch state stuff - we can now generate two identical hollow batches that get "combined" when they're put in the same set, and then only one of them is available to pair back up with the spine batches later. I think a zero-hollow-batch spinebatch avoids at least that problem...

Reply:
I'm not actually sure how to produce one of those? Here's what I'm running, which isn't trying to make any HollowBatches:

```
            let id = {
                let id = self.next_id;
                self.next_id += 1;
                SpineId(id, self.next_id)
            };
            let batch = SpineBatch::empty(
                id,
                self.upper.clone(),
                self.upper.clone(),
                Antichain::from_elem(T::minimum()),
            );
            let index = usize::try_from(effort).unwrap_or(0).next_power_of_two();
            self.introduce_batch(batch, usize::cast_from(index.trailing_zeros()), log);
```

Reply:
Oh I see, we probably just have to manually construct the `SpineBatch`

Reply:
Okay I typed that up and it seems to be a rabbit hole of lining up all our internal invariants for SpineIds, etc. Feeling a touch stuck here

Reply:
Ah yeah, extremely non-obvious that `SpineBatch::empty` would do this! (I'd certainly forgotten.)

I suspect the invariant situation gets easier to deal with once I've deleted the legacy spine-diff code, which should be real soon assuming last week's change behaves as expected. Maybe we start with only applying fuel for now, and then follow up with the empty batch work and strengthen the invariants once that's done? Whatever feels like it makes this less of a slog seems fine to me I think...

Reply:
Wfm! Done